File: agent/src/main/java/com/smancode/smanagent/tools/ToolResult.java
package com.smancode.smanagent.tools;

import java.time.Instant;
import java.util.List;
import java.util.Map;

/**
 * 工具执行结果
 */
public class ToolResult {

    /**
     * 是否成功
     */
    private boolean success;

    /**
     * 返回数据
     */
    private Object data;

    /**
     * 显示标题
     */
    private String displayTitle;

    /**
     * 显示内容
     */
    private String displayContent;

    /**
     * 错误信息
     */
    private String error;

    /**
     * 执行时长（毫秒）
     */
    private long executionTimeMs;

    /**
     * 执行时间
     */
    private Instant executionTime;

    /**
     * 相对路径（用于后续重新读取或修改）
     */
    private String relativePath;

    /**
     * 相关文件路径列表（用于 find_file 等工具）
     * <p>
     * 存储 find_file 找到的所有匹配文件路径
     */
    private List<String> relatedFilePaths;

    /**
     * 元数据（存储扩展信息）
     * <p>
     * 例如: absolutePath, totalLines, startLine, endLine 等
     */
    private Map<String, Object> metadata;

    public ToolResult() {
        this.executionTime = Instant.now();
    }

    public boolean isSuccess() {
        return success;
    }

    public void setSuccess(boolean success) {
        this.success = success;
    }

    public Object getData() {
        return data;
    }

    public void setData(Object data) {
        this.data = data;
    }

    public String getDisplayTitle() {
        return displayTitle;
    }

    public void setDisplayTitle(String displayTitle) {
        this.displayTitle = displayTitle;
    }

    public String getDisplayContent() {
        return displayContent;
    }

    public void setDisplayContent(String displayContent) {
        this.displayContent = displayContent;
    }

    public String getError() {
        return error;
    }

    public void setError(String error) {
        this.error = error;
    }

    public long getExecutionTimeMs() {
        return executionTimeMs;
    }

    public void setExecutionTimeMs(long executionTimeMs) {
        this.executionTimeMs = executionTimeMs;
    }

    /**
     * 创建成功结果
     *
     * @param data         数据
     * @param displayTitle 显示标题
     * @param displayContent 显示内容
     * @return 成功结果
     */
    public static ToolResult success(Object data, String displayTitle, String displayContent) {
        ToolResult result = new ToolResult();
        result.setSuccess(true);
        result.setData(data);
        result.setDisplayTitle(displayTitle);
        result.setDisplayContent(displayContent);
        return result;
    }

    /**
     * 创建失败结果
     *
     * @param error 错误信息
     * @return 失败结果
     */
    public static ToolResult failure(String error) {
        ToolResult result = new ToolResult();
        result.setSuccess(false);
        result.setError(error);
        return result;
    }

    /**
     * 计算执行时长
     */
    public void calculateDuration() {
        if (executionTime != null) {
            this.executionTimeMs = java.time.Duration.between(executionTime, Instant.now()).toMillis();
        }
    }

    public String getRelativePath() {
        return relativePath;
    }

    public void setRelativePath(String relativePath) {
        this.relativePath = relativePath;
    }

    public List<String> getRelatedFilePaths() {
        return relatedFilePaths;
    }

    public void setRelatedFilePaths(List<String> relatedFilePaths) {
        this.relatedFilePaths = relatedFilePaths;
    }

    public Map<String, Object> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, Object> metadata) {
        this.metadata = metadata;
    }

    /**
     * 创建成功结果（带相对路径）
     *
     * @param data         数据
     * @param relativePath 相对路径
     * @param displayContent 显示内容
     * @return 成功结果
     */
    public static ToolResult successWithPath(Object data, String relativePath, String displayContent) {
        ToolResult result = new ToolResult();
        result.setSuccess(true);
        result.setData(data);
        result.setRelativePath(relativePath);
        result.setDisplayTitle(relativePath);  // 标题使用相对路径
        result.setDisplayContent(displayContent);
        return result;
    }
}

File: src/main/kotlin/com/smancode/smanagent/tools/ToolResult.kt
package com.smancode.smanagent.tools

import java.time.Instant

/**
 * 工具执行结果
 */
class ToolResult {

    /**
     * 是否成功
     */
    var isSuccess: Boolean = false

    /**
     * 返回数据
     */
    var data: Any? = null

    /**
     * 显示标题
     */
    var displayTitle: String? = null

    /**
     * 显示内容
     */
    var displayContent: String? = null

    /**
     * 错误信息
     */
    var error: String? = null

    /**
     * 执行时长（毫秒）
     */
    var executionTimeMs: Long = 0

    /**
     * 执行时间
     */
    var executionTime: Instant = Instant.now()

    /**
     * 相对路径（用于后续重新读取或修改）
     */
    var relativePath: String? = null

    /**
     * 相关文件路径列表（用于 find_file 等工具）
     *
     * 存储 find_file 找到的所有匹配文件路径
     */
    var relatedFilePaths: List<String>? = null

    /**
     * 元数据（存储扩展信息）
     *
     * 例如: absolutePath, totalLines, startLine, endLine 等
     */
    var metadata: Map<String, Any>? = null

    /**
     * 批量执行的子工具结果列表
     *
     * 用于 batch 工具存储所有子工具的执行结果
     */
    var batchSubResults: List<BatchSubResult>? = null

    /**
     * 摘要（用于压缩后的结果）
     *
     * 存储工具结果的摘要，用于减少 Token 使用
     */
    var summary: String? = null

    /**
     * 计算执行时长
     */
    fun calculateDuration() {
        executionTimeMs = java.time.Duration.between(executionTime, Instant.now()).toMillis()
    }

    companion object {
        /**
         * 创建成功结果
         *
         * @param data         数据
         * @param displayTitle 显示标题
         * @param displayContent 显示内容
         * @return 成功结果
         */
        fun success(data: Any?, displayTitle: String?, displayContent: String?): ToolResult {
            return ToolResult().apply {
                isSuccess = true
                this.data = data
                this.displayTitle = displayTitle
                this.displayContent = displayContent
            }
        }

        /**
         * 创建失败结果
         *
         * @param error 错误信息
         * @return 失败结果
         */
        fun failure(error: String?): ToolResult {
            return ToolResult().apply {
                isSuccess = false
                this.error = error
            }
        }

        /**
         * 创建成功结果（带相对路径）
         *
         * @param data         数据
         * @param relativePath 相对路径
         * @param displayContent 显示内容
         * @return 成功结果
         */
        fun successWithPath(data: Any?, relativePath: String?, displayContent: String?): ToolResult {
            return ToolResult().apply {
                isSuccess = true
                this.data = data
                this.relativePath = relativePath
                this.displayTitle = relativePath  // 标题使用相对路径
                this.displayContent = displayContent
            }
        }
    }
}
